### ISSUE TITLE
attach meta info?

### ISSUE THREAD

**Issue by stephanos** [atteo/classindex#28]
Would it be feasible to attach meta information to each class in the index?

Here is our use case:
We have several hundred pages, they all inherit from BasePage.java. There is also a mapping file, PageIndex.java they is basically a big hash map that looks like this:
```java
   .put("/some/path", "com.myproject.pages.XYZ")
```

The reason we need to use the full qualified name is that we want to prevent the classloading of the page (slow on our platform). So in order to apply classindex to this we'd need to attach the page's path to the index (since we can't load the entire class and evaluate any runtime annotations).

So, basically I'd imagine a annotation like @IndexMeta("/some/path") and the ability to read it from the classindex API.

I know this is quite a big concept, but I just wanted to throw it out there :)

Stephan

**Comment by marembo2008** [comment]
I think the classindex does not need to add this.
You can define your own annotation like:
```
@Retention(RUNTIME)
@Target(TYPE)
@IndexAnnotated
public @interface PageIndex{
   String path();
}
```
And then at boot, in your post construct, or however you do it, you just process the pages

```
@PostConstruct
void initializePageIndeces(){
    final Iterable<Class<?>> pages = ClassIndex.getAnnotated(PageIndex.class);
    for(final Class<?> pageIndexClass:pages) {
        final PageIndex = pageIndexClass:pages.getAnnotation(PageIndex.class);
    }
}
```

**Comment by stephanos** [comment]
Well, no. I can't load it on boot since the classloading is slow (believe me, we hate it, too). And the whole point of this is to prevent any classloading ...

**Comment by stephanos** [comment]
The `ClassIndex.getAnnotated` loads all the classes. This is too slow on our platform :(

**Comment by marembo2008** [comment]
So i guess you would like to do something like:
```
final PageIndex pageIndex = getHoldOfPageIndexForPath("/some/path");
Class<? extends BasePage> pageClass = ClassIndex.getAnnotated(pageIndex);
```
So that you load the class at the point it is needed.
That would be a big toll, and the use case may actually be there, as i do this quite often in CDI.
I would try to tickle with it, and then see if i can actually add such a functionality. if you do before me, let me get an update

**Comment by stephanos** [comment]
Actually I want to do this:

```java
void initializePageIndeces(){
    HashMap<String, String> pathToPageClass = new HashMap<>();
    Iterable<String> pageClasses = ClassIndex.getSubclassesNames(BasePage.class);
    for (String pageClass : pageClasses) {
        String path = (String) ClassIndex.getSubclassMetaData(BasePage.class, pageClass);
        pathToPageClass.put(path, pageClass);
    }
}```

I should have written this in the first place :)
